R/driver-oracle.R

```diff
@@ -122,7 +122,20 @@ setMethod("odbcConnectionColumns_", c("Oracle", "character"),
       )
     }
 
-    dbGetQuery(conn, query)
+    res <- dbGetQuery(conn, query)
+
+    # SQLColumns does not correctly describe buffer sizes for date/time
+    # fields ( nor does our custom query above ).  We use these data
+    # as a fallback for SQLDescribeParam ( also incorrectly describes
+    # parameters for ORACLE ) when writing using prepared statements ( see
+    # `dbAppendTable` ).  Fix them here.  91 and 93 are SQL_TYPE_DATE,
+    # and SQL_TYPE_TIMESTAMP, respectively in sql.h.  Six and sixteen
+    # are the sizes of nanodbc::date and nanodbc::timestmp in bytes
+    # (#349, #350, #391).
+    res$data_type <- as.numeric(res$data_type)
+    res[res$field.type == "DATE", c("data_type", "column_size")] <- c(91, 6)
+    res[grepl("TIMESTAMP", res$field.type), c("data_type", "column_size")] <- c(93, 16)
```

Comment:
This is a very minor quibble but I don't trust this sort of cross-column subset-assignment, so I'd prefer to see something more like:

```R
res$data_type[res$field.type == "DATE"] <- 91
res$column_size[res$field.type == "DATE"] <- 6
```

I suspect that might also eliminate the need for `as.numeric(res$data_type)` above.

Reply:
Thanks.  Done - let me know if that's not what you had in mind.

I think the `as.numeric` may still be needed, but happy to update if you guys think there's a better way.

```
 ┃  Browse[1]> res <- dbGetQuery(conn, query)
 ┃  Browse[1]> str(res)
 ┃  'data.frame':   5 obs. of  17 variables:
 ┃   $ name                   : chr  "datetime" "date" "integer" "double" ...
 ┃   $ field.type             : chr  "TIMESTAMP(6)" "DATE" "NUMBER" "BINARY_DOUBLE" ...
 ┃   $ table_name             : chr  "test_table" "test_table" "test_table" "test_table" ...
 ┃   $ schema_name            : chr  "SA" "SA" "SA" "SA" ...
 ┃   $ catalog_name           : chr  NA NA NA NA ...
 ┃   $ data_type              : chr  NA NA NA NA ...
 ┃   $ column_size            : num  NA NA NA NA 255
 ┃   $ buffer_length          : num  11 16 40 8 255
 ┃   $ decimal_digits         : num  6 NA 0 NA NA
 ┃   $ numeric_precision_radix: chr  NA NA NA NA ...
 ┃   $ remarks                : chr  NA NA NA NA ...
 ┃   $ column_default         : chr  NA NA NA NA ...
 ┃   $ sql_data_type          : chr  NA NA NA NA ...
 ┃   $ sql_datetime_subtype   : chr  NA NA NA NA ...
 ┃   $ char_octet_length      : num  0 0 0 0 255
 ┃   $ ordinal_position       : num  1 2 3 4 5
 ┃   $ nullable               : num  1 1 1 1 1
 ┃  Browse[1]>     isDate <- res$field.type == "DATE"
 ┃      res$data_type[isDate] <- 91
 ┃      res$column_size[isDate] <- 6
 ┃      isTimestamp <- grepl("TIMESTAMP", res$field.type)
 ┃      res$data_type[isTimestamp] <- 93
 ┃      res$column_size[isTimestamp] <- 16
 ┃  Browse[1]> str(res)
 ┃  'data.frame':   5 obs. of  17 variables:
 ┃   $ name                   : chr  "datetime" "date" "integer" "double" ...
 ┃   $ field.type             : chr  "TIMESTAMP(6)" "DATE" "NUMBER" "BINARY_DOUBLE" ...
 ┃   $ table_name             : chr  "test_table" "test_table" "test_table" "test_table" ...
 ┃   $ schema_name            : chr  "SA" "SA" "SA" "SA" ...
 ┃   $ catalog_name           : chr  NA NA NA NA ...
 ┃   $ data_type              : chr  "93" "91" NA NA ...
 ┃   $ column_size            : num  16 6 NA NA 255
 ┃   $ buffer_length          : num  11 16 40 8 255
 ┃   $ decimal_digits         : num  6 NA 0 NA NA
 ┃   $ numeric_precision_radix: chr  NA NA NA NA ...
 ┃   $ remarks                : chr  NA NA NA NA ...
 ┃   $ column_default         : chr  NA NA NA NA ...
 ┃   $ sql_data_type          : chr  NA NA NA NA ...
 ┃   $ sql_datetime_subtype   : chr  NA NA NA NA ...
 ┃   $ char_octet_length      : num  0 0 0 0 255
 ┃   $ ordinal_position       : num  1 2 3 4 5
 ┃   $ nullable               : num  1 1 1 1 1
```
This is a bit of a function of the silly way the `data_type` column is formulated in the query ( always `NULL` ) - we can probably do something better to make sure that the column comes back as numeric, but there is also probably some value in keeping the query as close as the original `SQLColumns` implementation for the OEM oracle driver.